### PR TITLE
Improve decorators

### PR DESCRIPTION
**Pull request by jacobherrington**
I'm not completely confident that testing would catch all of the problems that could arise from this, but it is passing tests...

Open to feedback on whether or not this change is worth it, I think it is.

### CODE REVIEW

File: lib/spree/auth/engine.rb [hunk 1]
```diff
@@ -32,8 +32,8 @@ def self.prepare_backend
           lib/assets/javascripts/spree/backend/solidus_auth.css
         ]
 
-        Dir.glob(File.join(File.dirname(__FILE__), "../../controllers/backend/*/*/*_decorator*.rb")) do |c|
-          Rails.configuration.cache_classes ? require(c) : load(c)
+        Dir.glob(File.join(File.dirname(__FILE__), "../../decorators/**/*.rb")) do |c|
+          require_dependency(c)
```
Comment (spaghetticode) [inline lib/spree/auth/engine.rb]: I think this change removes the reloading of these classes in development, a feature that I tend to consider useful. Is there a reason for removing the feature?

Comment (jacobherrington) [reply lib/spree/auth/engine.rb]: That's a good point, I actually see another (fairly obvious) issue with this now that you bring that up. :) Thanks!

Comment (kennyadsl) [reply lib/spree/auth/engine.rb]: I don't think this removes the class reloading in development, `required_dependency` should take care of that.

Comment (spaghetticode) [reply lib/spree/auth/engine.rb]: @kennyadsl you're right, `require_dependency` is not the problem. The problem lies in the file structure that does not match the constant naming, so they don't get correctly reloaded.

For example reloading would work with the constant
```ruby
module Spree
  module Controllers
    module Frontend
      module CheckoutControllerDecorator
      end
    end
  end
end
```
defined in `app/decorators/spree/controllers/frontend/checkout_controller_decorator.rb`

Comment (jacobherrington) [reply lib/spree/auth/engine.rb]: @spaghetticode Can you take another look? I changed the namespacing, but I'm not sure it's ideal, thoughts?

Comment (spaghetticode) [reply lib/spree/auth/engine.rb]: @jacobherrington it seems to me the reload is not working yet. I was able to make it work by adding the  `lib` directory to the `autoload_paths`:

```ruby
    class Engine < Rails::Engine
      isolate_namespace Spree
      engine_name 'solidus_auth'

      config.autoload_paths <<  File.expand_path("../../..", __FILE__)
```

Regarding the namespacing and directory structure, if we were to decorate a Spree model later, where would we put this decorator?
I'm thinking a structure like the following one, by mimicking rails `app` folder,  may look a little more familiar:
```
decorators
├── controllers
│   └── spree
│       ├── backend
│       │   └── admin_controller_decorator.rb
│       └── frontend
│           └── checkout_controller_decorator.rb
└── models
    └── spree

```
The existing files, except for being moved to a different structure, would remain unchanged. 

More in general, as most of the extensions are still relying on `Module#class_eval`, transitioning to `Module#prepend` will probably raise this kind of questions also in their regards. I'm thinking that maybe we should take some time to decide on conventional namespacing and decorators directory structure 🤔

Comment (kennyadsl) [reply lib/spree/auth/engine.rb]: I agree with @spaghetticode, we should create a new convention before changing working extensions, or we could end up making change twice making users life harder. A discussion started [here](https://github.com/solidusio/solidus/issues/3010) and I think we should finalize a decision there.

**Comment by jacobherrington** [comment]
Based on the discussion with @spaghetticode and @kennyadsl, I've decided to use this PR just for removing class_eval from the existing decorators.

The decision on a new pattern for organizing "decorators" in Solidus should be discussed and settled in [the issue on the main repo](https://github.com/solidusio/solidus/issues/3010).

File: lib/spree/auth/engine.rb (hunk 1 above)
Comment (jacobherrington) [reply lib/spree/auth/engine.rb]: I have decided to just stick to removing the usage of `class_eval` in this extension. We can make this change later after a decision has been made, good call!

File: lib/controllers/backend/spree/admin/admin_controller_decorator.rb [hunk 2]
```diff
@@ -1,14 +1,20 @@
 # frozen_string_literal: true
 
-Spree::Admin::BaseController.class_eval do
```
Comment (kennyadsl) [inline lib/controllers/backend/spree/admin/admin_controller_decorator.rb]: file should be named `base_controller_decorato.rb` right?

Comment (jacobherrington) [reply lib/controllers/backend/spree/admin/admin_controller_decorator.rb]: I actually have seen this pattern in several places, but I agree with you. Will make that change.

**Pull request merged**